### PR TITLE
ref(reader): Return date and datetimes instead of strings

### PR DESCRIPTION
**Pull request by tkaemming**
This code will need to be shared between the HTTP (JSON) reader (getsentry/snuba#406) and the native reader, so this extracts it from the `NativeDriverReader`.

Formatting the datetime is object is already handled as part of the JSON serializer in the API: https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/api.py#L156

It makes more sense to pass around a datetime object than a string, since this avoids unnecessary type conversions in components such as the splitter.

### CODE REVIEW

File: snuba/reader.py [hunk 1]
```diff
@@ -36,11 +37,28 @@ def execute(
         raise NotImplementedError
 
 
-class NativeDriverReader(Reader):
+DATE_TYPE_RE = re.compile(r"(Nullable\()?Date\b")
+DATETIME_TYPE_RE = re.compile(r"(Nullable\()?DateTime\b")
+
+
+def transform_date_columns(result: Result) -> Result:
```
Comment (manuzope) [inline snuba/reader.py]: How do you feel about having some tests for this function?

Comment (tkaemming) [reply snuba/reader.py]: There should be general coverage via the API tests, but a unit test wouldn't be a bad thing either. At the same time, this is all pretty likely going to go through another round of changes in the near future so I'm hesitant to invest too much time in it.

**Comment by tkaemming** [comment]
Out of scope for this change, but there is also a bug here:

```python
>>> import pprint
>>> from snuba import settings
>>> from snuba.api import clickhouse_ro
>>> from snuba.reader import NativeDriverReader
>>> 
>>> native = NativeDriverReader(clickhouse_ro)
>>> pprint.pprint(native.execute('SELECT now() FORMAT JSON'))
{'data': [{'now()': datetime.datetime(2019, 8, 14, 18, 22, 54, tzinfo=tzutc())}],
 'meta': [{'name': 'now()', 'type': 'DateTime'}]}
>>> pprint.pprint(native.execute('SELECT [now()] FORMAT JSON'))
{'data': [{'array(now())': (datetime.datetime(2019, 8, 14, 18, 22, 54),)}],
 'meta': [{'name': 'array(now())', 'type': 'Array(DateTime)'}]}
```

File: snuba/reader.py (hunk 1 above)
Comment (manuzope) [reply snuba/reader.py]: Fair :+1:

**Review by manuzope — APPROVED** [review]

File: snuba/reader.py (hunk 1 above)
Comment (tkaemming) [reply snuba/reader.py]: There will probably be a change soon to change the type of `Result`, I promise to add them at that point (and fix [the bug with non-scalar values noted here](https://github.com/getsentry/snuba/pull/414#issuecomment-521361553).)

I'm putting this in writing so you can hold me to it.  🙂

**Pull request merged**